### PR TITLE
Covalent Cloud GPU tutorial

### PR DESCRIPTION
**Pull request by kessler-frost**
------------------------------------------------------------------------------------------------------------

**Title:** Running GPU-accelerated quantum circuit simulations on Covalent Cloud using PennyLane

**Summary:**

**Relevant references:**

**Possible Drawbacks:**

**Related GitHub Issues:**

----
If you are writing a demonstration, please answer these questions to facilitate the marketing process.

* GOALS — Why are we working on this now?

  *Eg. Promote a new PL feature or show a PL implementation of a recent paper.*

* AUDIENCE — Who is this for?

  *Eg. Chemistry researchers, PL educators, beginners in quantum computing.*

* KEYWORDS — What words should be included in the marketing post?

* Which of the following types of documentation is most similar to your file? 
(more details [here](https://www.notion.so/xanaduai/Different-kinds-of-documentation-69200645fe59442991c71f9e7d8a77f8))
    
- [ ] Tutorial
- [ ] Demo
- [ ] How-to

### CODE REVIEW

**Comment by kessler-frost** [comment]
One issue I had when converting this from a jupyter notebook to a `.py` file using `notebook_to_demo.py` was this error:
```python
❯ python notebook_converter/notebook_to_demo.py covalent_cloud_gpu.ipynb --is-executable=False
Traceback (most recent call last):
  File "/Users/neptune/dev/qml/notebook_converter/notebook_to_demo.py", line 360, in <module>
    nb_py = convert_notebook_to_python(
  File "/Users/neptune/dev/qml/notebook_converter/notebook_to_demo.py", line 262, in convert_notebook_to_python
    cell_id = cell["id"]
KeyError: 'id'
```

I edited [Line 262](https://github.com/PennyLaneAI/qml/blob/f9df21d666075ea6998894a8c63bc96641437cba/notebook_converter/notebook_to_demo.py#L262) to `cell_id = i` instead and then was able to run the conversion script fine.

**Comment by kessler-frost** [comment]
Thanks a lot @ikurecic ! Updated the PR ✅ . Hopefully I haven't missed anything else 🤞 .

>You can also transfer the included plot into the demonstrations_assets folder by following the instructions [here](https://github.com/PennyLaneAI/qml) with the syntax below.

My bad actually I must've missed the right dir when converting it from `.ipynb`, updated ✅ .

File: demonstrations/covalent_cloud_gpu.metadata.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "basedOnPapers": [],
+    "referencedByPapers": [],
+    "relatedContent": [
+        {
+            "type": "demonstration",
+            "id": "tutorial_univariate_qvr",
+            "weight": 1.0
+        }
+    ],
+    "hardware": []
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.metadata.json:40]:
```suggestion
    "hardware": [
        {
            "id": "covalent",
            "link": "https://docs.covalent.xyz/docs/cloud/tutorials-cloud/pennylane_tutorial/",
            "logo": "/_static/hardware_logos/covalent.png"
        }
    ]
```

This suggestion can be applied after https://github.com/PennyLaneAI/qml/pull/1112 is merged in (cc @ikurecic)

File: demonstrations/covalent_cloud_gpu.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
+r"""Running GPU-accelerated quantum circuit simulations on Covalent Cloud using PennyLane
+===================================================================
+"""
+
+##############################################################################
+
+# In this tutorial, we’ll demonstrate how to run GPU-accelerated quantum circuit simulations
+# on `Covalent Cloud <https://www.covalent.xyz>`__ using PennyLane. We will focus on a specific example around
+# quantum SVM to demonstrate how easy  it is to  run GPU accelerated quantum circuit simulation on Covalent Cloud.
+
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:10]:
```suggestion
#
```

File: demonstrations/covalent_cloud_gpu.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
+r"""Running GPU-accelerated quantum circuit simulations on Covalent Cloud using PennyLane
+===================================================================
+"""
+
+##############################################################################
+
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:6]:
```suggestion
#
```

File: demonstrations/covalent_cloud_gpu.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+#    - cudatoolkit>=11.8
+#    - pip:
+#      - cuquantum==23.10.0
+#      - matplotlib==3.8.2
+#      - Pennylane==0.34.0
+#      - PennyLane-Lightning[GPU]==0.34.0
+#      - scikit-learn==1.3.1
+#      - torch==2.1.2
+#      - covalent-cloud
+#    name: pennylane-gpu
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:93]: I believe all code blocks underneath `.. rst-class :: sphx-glr-script-out` need to be indented by 2 space to display correctly :)

Comment (josh146) [reply demonstrations/covalent_cloud_gpu.py:93]:
```suggestion
#  .. code-block: none
#
#     Waiting for environment pennylane-gpu to be ready..................
#     Name: pennylane-gpu
#     Status: READY
#     Estimated Time: 1002 seconds
#     Notes:
#     	pip was added to the dependencies.
#     	Python version 3.10 was added to the dependencies.
#     Environment file contains:
#     ==========================
#     channels:
#     - conda-forge
#     dependencies:
#     - python=3.10
#     - pip
#     - cudatoolkit>=11.8
#     - pip:
#       - cuquantum==23.10.0
#       - matplotlib==3.8.2
#       - Pennylane==0.34.0
#       - PennyLane-Lightning[GPU]==0.34.0
#       - scikit-learn==1.3.1
#       - torch==2.1.2
#       - covalent-cloud
#     name: pennylane-gpu
```

Comment (josh146) [reply demonstrations/covalent_cloud_gpu.py:93]: (same will need to be done for other output cells)

File: demonstrations/covalent_cloud_gpu.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+dispatch_id = cc.dispatch(run_qsvm)(n_samples=64, test_size=0.2)
+print("Dispatch ID:", dispatch_id)
+
+##############################################################################
+# .. rst-class :: sphx-glr-script-out
+#
+# .. code-block: none
+#
+#    Dispatch ID: 0b5d3a08-fe9c-4dc2-910b-0be6eb925663
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:214]:
```suggestion
#  .. code-block: none
#
#     Dispatch ID: 0b5d3a08-fe9c-4dc2-910b-0be6eb925663
```

File: demonstrations/covalent_cloud_gpu.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+#
+#    Train accuracy: 64.7%
+#    Test accuracy: 76.9%# .. rst-class :: sphx-glr-script-out
+#
+# .. code-block: none
+#
+#
+# .. figure:: ../_static/demonstration_assets/covalent_cloud_gpu/covalent_cloud_gpu_19_1.png
+#    :align: center
+#    :width: 90%
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:242]:
```suggestion
# .. rst-class :: sphx-glr-script-out
#
#  .. code-block: none
#
#     Train accuracy: 64.7%
#     Test accuracy: 76.9%
#
# .. figure:: ../_static/demonstration_assets/covalent_cloud_gpu/covalent_cloud_gpu_19_1.png
#    :align: center
#    :width: 90%
```

File: demonstrations/covalent_cloud_gpu.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+# Conclusion
+# ==========
+#
+# In this tutorial, we demonstrated how to run quantum circuit simulations on GPUs via Covalent Cloud.
+# We used Pennylane to define a simple quantum kernel, and then trained and tested a QSVM on a
+# 2-dimensional dataset. To make the most of this tutorial, try experimenting with different
+# datasets/kernels or increasing the dataset dimension, to gain a greater advantage from GPU
+# acceleration.
+#
+# The cost of running this workflow is approximately $0.27. The full code is available below.
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:254]:
```suggestion
# The cost of running this workflow is approximately $0.27$. The full code is available below.
```

**Comment by kessler-frost** [comment]
Thanks @josh146 !

File: demonstrations/covalent_cloud_gpu.py (hunk 7 above)
Comment (josh146) [reply demonstrations/covalent_cloud_gpu.py]: Oh, oops -- I thought this was meant to be math mode 😆 In that case, you want to make it 

```rest
\$0.27.
```

Comment (kessler-frost) [reply demonstrations/covalent_cloud_gpu.py]: 😄 no worries, done ✅ .

File: demonstrations/covalent_cloud_gpu.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+#     Train accuracy: 64.7%
+#     Test accuracy: 76.9%
+#
+# .. figure:: ../_static/demonstration_assets/covalent_cloud_gpu/covalent_cloud_gpu_19_1.png
+#    :align: center
+#    :width: 90%
+
+##############################################################################
+# Conclusion
+# ==========
```
Comment (josh146) [inline demonstrations/covalent_cloud_gpu.py:243]:
```suggestion
# Conclusion
# ----------
```

(since `=` is used for the page title)

**Comment by josh146** [comment]
Hey @kessler-frost, would you be able to give us [permissions to push](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to this branch :) That way we can help out with some of the operations here.

**Comment by wjcunningham7** [comment]
> Hey @kessler-frost, would you be able to give us [permissions to push](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to this branch :) That way we can help out with some of the operations here.

You should have access now @josh146

**Comment by josh146** [comment]
Thank you!

**Comment by josh146** [comment]
@wjcunningham7 thank you!

**Comment by josh146** [comment]
Hey all! There was a slight accident when I tried to merge the master branch of the QML repo into this PR --- I think because the forked repo did not create a new branch (and instead used `master` as the branch), the sync caused this PR to be accidentally merged.

I've reverted the merge, and everything is find, but it doesn't look like I am able to re-open this PR. To try and fix this, I create a new PR, this time with a new branch name (not `master`), here: https://github.com/PennyLaneAI/qml/pull/1115/

@kessler-frost you should have full access to push to this PR.

Apologies!

**Comment by kessler-frost** [comment]
Thanks @josh146 !

**Pull request merged**